File: ui/src/app/app.constants.ts
export const AppConstants = {
  // An id designating a new record is to be created.
  NEW_RECORD_ID: "new",

  // The date format to use.
  DATE_FORMAT: "yyyy-MM-dd HH:mm:ss",

  // The different types of security keys and CERTIFICATE.
  KEY_TYPE: {
    PRIVATE_KEY: "PRIVATE",
    PUBLIC_KEY: "PUBLIC",
    CERTIFICATE: "CERTIFICATE"
  },

  // A list of application-wide settings (see also the counterpart Java class AppConstants.java).
  NAMED_SETTING: {
    DEVICE_PROVISIONING_URL: "DEVICE_PROVISIONING_URL",
    DEVICE_PROVISIONING_SECURE: "DEVICE_PROVISIONING_SECURE",
    DEVICE_REGISTRATION_MODE: "DEVICE_REGISTRATION_MODE",
    DEVICE_REGISTRATION_SECRET: "DEVICE_REGISTRATION_SECRET",
    DEVICE_ROOT_CA: "DEVICE_ROOT_CA",
    PLATFORM_CERTIFICATE: "PLATFORM_CERTIFICATE",
    KAFKA_TOPIC_COMMAND_REQUEST: "KAFKA_TOPIC_COMMAND_REQUEST",
    DEVICE_PROVISIONING_CACHE_TIME: "DEVICE_PROVISIONING_CACHE_TIME",
    DEVICE_PUSHED_TAGS: "DEVICE_PUSHED_TAGS"
  },

  // Keystore items.
  KEYSTORE: {
    ITEM: {
      RESOURCE_TYPE: {
        DEVICE: "DEVICE",
        CERTIFICATE: "CERT",
        CA: "CA",
        TAG: "TAG"
      },
      KEY_TYPE: {
        PRIVATE_KEY: "PRIVATE",
        CERTIFICATE: "CERT"
      }
    }
  },

  // Device settings.
  DEVICE: {
    ATTRIBUTE: {
      TYPE: {
        STRING: "STRING",
        BOOLEAN: "BOOLEAN",
        BYTE: "BYTE",
        SHORT: "SHORT",
        INTEGER: "INTEGER",
        LONG: "LONG",
        BIG_DECIMAL: "BIG_DECIMAL",
        UNKNOWN: "UNKNOWN"
      }
    },
    TYPE: {
      ESTHESIS: "ESTHESIS"
    },
    STATUS: {
      DISABLED: "DISABLED",
      PREREGISTERED: "PREREGISTERED",
      REGISTERED: "REGISTERED",
      APPROVAL: "APPROVAL"
    },

    SETTING: {
      DEVICE_GEO_LAT: "DEVICE_GEO_LAT",
      DEVICE_GEO_LON: "DEVICE_GEO_LON"
    },

    COMMAND: {
      TYPE: {
        EXECUTE: "e",
        HEALTH: "h",
        PING: "p",
        FIRMWARE: "f",
        REBOOT: "r",
        SHUTDOWN: "s"
      },
      EXECUTION: {
        SYNCHRONOUS: "s",
        ASYNCHRONOUS: "a"
      }
    }
  },

  // User statuses.
  USER_STATUS: {
    DISABLED: 0,
    ENABLED: 1
  },

  // Provisioning.
  PROVISIONING: {
    TYPE: {
      ESTHESIS: "ESTHESIS",
      WEB: "WEB",
      FTP: "FTP",
      MINIO: "MINIO"
    },
    CONFIG: {
      FTP: {
        HOST: "HOST",
        PORT: "PORT",
        USERNAME: "USERNAME",
        PASSWORD: "PASSWORD",
        PATH: "PATH",
        PASSIVE: "PASSIVE"
      }
    },
    CACHE_STATUS: {
      NOT_STARTED: "NOT_STARTED",
      IN_PROGRESS: "IN_PROGRESS",
      COMPLETED: "COMPLETED",
      FAILED: "FAILED"
    }
  },

  // Campaign constants.
  CAMPAIGN: {
    STATE: {
      CREATED: "CREATED",
      RUNNING: "RUNNING",
      PAUSED_BY_USER: "PAUSED_BY_USER",
      PAUSED_BY_WORKFLOW: "PAUSED_BY_WORKFLOW",
      TERMINATED_BY_WORKFLOW: "TERMINATED_BY_WORKFLOW",
      TERMINATED_BY_USER: "TERMINATED_BY_USER",
    },
    TYPE: {
      PROVISIONING: "PROVISIONING",
      EXECUTE_COMMAND: "EXECUTE_COMMAND",
      REBOOT: "REBOOT",
      SHUTDOWN: "SHUTDOWN",
      PING: "PING",
    },
    CONDITION: {
      TYPE: {
        DATETIME: "DATETIME",
        SUCCESS: "SUCCESS",
        PROPERTY: "PROPERTY",
        PAUSE: "PAUSE",
        BATCH: "BATCH",
      },
      STAGE: {
        ENTRY: "ENTRY",
        EXIT: "EXIT",
        INSIDE: "INSIDE"
      },
      OP: {
        BEFORE: "BEFORE",
        AFTER: "AFTER",
        FOREVER: "FOREVER",
        TIMER_MINUTES: "TIMER_MINUTES",
        EQUAL: "EQUAL",
        GT: "GT",
        LT: "LT",
        GTE: "GTE",
        LTE: "LTE",
      }
    },
    MEMBER_TYPE: {
      DEVICE: "DEVICE",
      TAG: "TAG",
    }
  },

  DIALOG_RESULT: {
    CANCEL: 0,
    SAVE: 1,
    DELETE: 2,
    OK: 3
  },

};

File: esthesis-core-backend/common/src/main/java/esthesis/common/AppConstants.java
package esthesis.common;

public class AppConstants {

	// The topic prefix to use when a Kafka topic needs to be created.
	//TODO to be removed with Settings preferences - is it used?
	public static final String KAFKA_TOPIC_PREFIX = "esthesis-";

	// A keyword used to indicate a new record being created.
	//TODO is it used?
	public static final String NEW_RECORD_ID = "new";

	// Redis key suffixes for measurements.
	public static final String REDIS_KEY_SUFFIX_VALUE_TYPE = "valueType";
	public static final String REDIS_KEY_SUFFIX_TIMESTAMP = "timestamp";
	public static final String REDIS_KEY_PROVISIONING_PACKAGE_FILE = "file";

	// Naming convention for hardware ids.
	public static final String HARDWARE_ID_REGEX = "^[a-zA-Z0-9_-]*$";

	// The size limit when displaying possibly large content in the logs.
	public static final int MESSAGE_LOG_ABBREVIATION_LENGTH = 4096;

	// The header containing the shared secret for device registration (when the platform operates
	// in that mode).
	//TODO where/how is this used?
	public static final String REGISTRATION_SECRET_HEADER_NAME = "X-ESTHESIS-REGISTRATION-SECRET";

	// Security roles.
	public static final String ROLE_USER = "user";
	public static final String ROLE_SYSTEM = "system";

	// Settings keys.
	public enum NamedSetting {
		SECURITY_ASYMMETRIC_KEY_SIZE, SECURITY_ASYMMETRIC_KEY_ALGORITHM,
		SECURITY_ASYMMETRIC_SIGNATURE_ALGORITHM,

		DEVICE_ROOT_CA, DEVICE_PROVISIONING_URL, DEVICE_REGISTRATION_MODE,
		DEVICE_REGISTRATION_SECRET,
		DEVICE_TAGS_ALGORITHM, DEVICE_GEO_LAT, DEVICE_GEO_LON,
		DEVICE_PROVISIONING_SECURE, DEVICE_PROVISIONING_CACHE_TIME,
		DEVICE_PUSHED_TAGS,

		KAFKA_TOPIC_COMMAND_REQUEST, KAFKA_TOPIC_TELEMETRY, KAFKA_TOPIC_METADATA,
		KAFKA_TOPIC_APPLICATION,
		// Application events
		//TODO we need a different naming convention, not starting with KAFKA_TOPIC
		KAFKA_TOPIC_EVENT_TAG_DELETE
	}

	public enum KeyType {
		PUBLIC, PRIVATE, CERTIFICATE
	}

	public enum DeviceRegistrationMode {
		DISABLED, OPEN, OPEN_WITH_SECRET, ID
	}

	public static class Device {

		public enum Type {
			ESTHESIS, OTHER
		}

		public enum Status {
			DISABLED, PREREGISTERED, REGISTERED
		}

		public static class Attribute {

			public enum Type {
				STRING, BOOLEAN, BYTE, SHORT, INTEGER, LONG, BIG_DECIMAL, UNKNOWN
			}
		}

	}

	public static class Keystore {

		public static class Item {

			public enum ResourceType {
				DEVICE, CERT, CA, TAG
			}

			public enum KeyType {
				PRIVATE, CERT
			}
		}
	}

	public static class Provisioning {

		public enum ConfigOption {
			FTP_HOST, FTP_PORT, FTP_USERNAME, FTP_PASSWORD, FTP_PATH, FTP_PASSIVE,
			WEB_URL, WEB_USERNAME, WEB_PASSWORD,
			MINIO_URL, MINIO_BUCKET, MINIO_OBJECT, MINIO_ACCESS_KEY, MINIO_SECRET_KEY,
			S3_OBJECT, S3_BUCKET
		}

		public enum Type {
			ESTHESIS, WEB, FTP, MINIO
		}

		public enum CacheStatus {
			NOT_STARTED, IN_PROGRESS, COMPLETED, FAILED
		}

		public static class Redis {

			public static final String DOWNLOAD_TOKEN_PACKAGE_ID = "DTPI";
			public static final String DOWNLOAD_TOKEN_CREATED_ON = "DTCO";

			private Redis() {
			}
		}
	}

	public static class Security {

		public enum Category {
			ABOUT,
			APPLICATION,
			AUDIT,
			CA,
			CAMPAIGN,
			CERTIFICATES,
			COMMAND,
			CRYPTO,
			DASHBOARD,
			DATAFLOW,
			DEVICE,
			INFRASTRUCTURE,
			KEYSTORE,
			KUBERNETES,
			PROVISIONING,
			SETTINGS,
			TAG,
			USERS,
			GROUPS,
			ROLES,
			POLICIES,
			NULL
		}

		public enum Operation {
			CREATE, READ, WRITE, DELETE, OTHER
		}

		public enum Permission {

			ALLOW, DENY;
		}

		public static class Ern {

			public static final String ROOT = "ern";
			public static final String SYSTEM = "esthesis";
			public static final String SUBSYSTEM = "core";
		}
	}

	public static class Campaign {

		public enum State {

			CREATED, RUNNING, PAUSED_BY_USER, PAUSED_BY_WORKFLOW,
			TERMINATED_BY_WORKFLOW, TERMINATED_BY_USER;
		}

		public enum Type {

			PROVISIONING, EXECUTE_COMMAND, REBOOT, SHUTDOWN, PING;
		}

		public static class Member {

			public enum Type {
				DEVICE, TAG;
			}
		}

		public static class Condition {

			public enum Type {
				DATETIME, SUCCESS, PROPERTY, PAUSE, BATCH;
			}

			public enum Stage {
				ENTRY, EXIT, INSIDE;
			}

			public enum Op {

				BEFORE, AFTER, ABOVE, BELOW, FOREVER, TIMER_MINUTES, EQUAL,
				GT, LT, GTE, LTE;
			}
		}
	}

}
